models/media_stats.sql

```diff
@@ -0,0 +1,126 @@
+{{ 
+  config(
+    materialized='table',
+  )
+}}
+
+with prep as (
+
+ 	select 
+    p.media_id,
+    p.duration,
+    p.title,
+    p.media_type,
+    p.media_player_type,
+    sum(p.play_time_sec) / 3600 as play_time_h,
+    avg(p.play_time_sec / 60) as avg_play_time_min,
+    sum(case when p.is_complete_play then 1 else 0 end) as complete_plays,
+    count(distinct p.page_view_id) as views,
+    count(distinct p.domain_sessionid) as "sessions",
+    avg(coalesce(p.play_time_sec / nullif(p.duration, 0), 0)) as retention_rate,
+    count(distinct p.domain_userid) as unique_users
+
+  from {{ ref("plays_by_pageview") }} p
+ 
+  group by 1, 2, 3, 4, 5
+
+)
+
+, valid_play_stats as (
+
+    select
+      media_id,
+      min(start_tstamp) as first_play,
+      max(start_tstamp) as last_play,
+      count(*) as valid_plays
+    
+    from {{ ref("plays_by_pageview") }}
+
+    where is_valid_play
+
+    group by 1
+
+)
+
+, impressions as (
+
+    select 
+      media_id,
+      count(*) as impressions
+    
+    from {{ ref("interactions") }}
+
+    where event_type = 'ready'
```

Comment:
Do you think we should include impressions in the `plays_by_pageview` model? I know strictly speaking it wouldnt be just plays then but just thinking out loud

Reply:
As we agreed on the group discussion, I have changed the logic to fit the impressions into `plays_by_pageview`. This has led me to realise I need to reconsider KPIs for `media_stats` in general. Currently I only have a rough idea which metric needs which logic (or more than one) out of the following 4 levels of filtering: let's say for `unique users` as a base we could take 1) all interactions including those cases when there is only a `ready` event within the page view so no actual interaction with the video 2) all page views where some interaction with the video happened but not necessarily play event 3) interactions within page views with SOME play_time_sec 4) interactions with valid plays (currently longer than 30 min). You will see I opted for calculating both 3) `unique_users_with_plays` and 4) `unique_users_with_valid_plays` whereas when I look at completion_rate, I only consider option 3) (`completion_rate_by_play`)